File: generated_code/GeminiPro/Java/030_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the property that
 * recently accessed elements are quick to access again. It performs basic
 * operations such as insertion, deletion, and search in O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Private Helper Methods for Splay and Rotations ---

    /**
     * Performs a right rotation on the subtree rooted at node y.
     * @param y The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node rotateRight(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node rotateLeft(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The splay operation moves the node with the given key (or the last accessed
     * node if the key is not found) to the root of the tree. This method uses a
     * top-down, iterative approach.
     * @param key The key of the node to splay to the root.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Use a dummy node to simplify linking of left and right subtrees.
        Node dummy = new Node(0);
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) break;
                // Zig-Zig (left-left)
                if (key < current.left.key) {
                    current = rotateRight(current);
                    if (current.left == null) break;
                }
                // Link current to the right tree (nodes > key)
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) break;
                // Zig-Zig (right-right)
                if (key > current.right.key) {
                    current = rotateLeft(current);
                    if (current.right == null) break;
                }
                // Link current to the left tree (nodes < key)
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = dummy.right;
        current.right = dummy.left;
        
        root = current;
    }

    // --- Public API Methods ---

    /**
     * Inserts a key into the splay tree. If the key already exists, the tree
     * is splayed on that key, but no new node is inserted.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        splay(key);

        // If key is already present, do nothing (tree is already splayed)
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree. If the key is not in the tree,
     * the tree is splayed on the last accessed node and no deletion occurs.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        splay(key);

        // If key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Key is at the root, now remove it
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Splay the largest element in the left subtree to its root.
            // This is achieved by splaying for the deleted key (which is guaranteed
            // to be larger than any key in the left subtree), causing the
            // maximum element to rise to the root of the left subtree.
            root = leftSubtree;
            splay(key);
            
            // Attach the original right subtree
            root.right = rightSubtree;
        }
    }

    /**
     * Searches for a key in the tree and splays the accessed node to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise {@code null}.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        if (root.key == key) {
            return key;
        }
        return null;
    }
}

File: generated_code/GeminiPro/Python/030_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class implements a dictionary-like set for storing unique integer keys.
    It supports insert, delete, and search operations. The key feature of a
    splay tree is that recently accessed elements are moved to the root of the
    tree, which can improve performance for certain access patterns (locality
    of reference).
    """

    class _Node:
        """A private class representing a node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _right_rotate(self, node):
        """Performs a right rotation on the given node."""
        new_root = node.left
        node.left = new_root.right
        new_root.right = node
        return new_root

    def _left_rotate(self, node):
        """Performs a left rotation on the given node."""
        new_root = node.right
        node.right = new_root.left
        new_root.left = node
        return new_root

    def _splay(self, key):
        """
        Splays the node with the given key to the root.

        If the key is not in the tree, the last accessed node on the search
        path is splayed to the root. This is a top-down splay implementation.
        """
        if not self.root or self.root.key == key:
            return

        # Use a dummy node to simplify linking of Left and Right subtrees
        dummy = self._Node(None)
        left_tree_max = dummy
        right_tree_min = dummy
        current = self.root

        while True:
            if key < current.key:
                if not current.left:
                    break
                # Zig-Zig case (left-left)
                if key < current.left.key:
                    current = self._right_rotate(current)
                    if not current.left:
                        break
                # Link current to the right tree (nodes larger than key)
                right_tree_min.left = current
                right_tree_min = current
                current = current.left
            elif key > current.key:
                if not current.right:
                    break
                # Zig-Zig case (right-right)
                if key > current.right.key:
                    current = self._left_rotate(current)
                    if not current.right:
                        break
                # Link current to the left tree (nodes smaller than key)
                left_tree_max.right = current
                left_tree_max = current
                current = current.right
            else:  # key == current.key
                break

        # Reassemble the tree
        left_tree_max.right = current.left
        right_tree_min.left = current.right
        current.left = dummy.right
        current.right = dummy.left
        self.root = current

    def insert(self, key):
        """
        Inserts an integer key into the tree.

        If the key already exists, it is splayed to the root.
        Otherwise, the new key is inserted and becomes the new root.
        
        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        self._splay(key)

        # If key is already present, splay has moved it to the root.
        if self.root.key == key:
            return

        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
        
        self.root = new_node

    def search(self, key):
        """
        Searches for a key in the tree and splays the accessed node.

        The accessed node (or its parent if the node is not found) is
        moved to the root of the tree.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False
        self._splay(key)
        return self.root.key == key

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is found, it is deleted, and the tree is restructured.
        If the key is not found, the tree is splayed on the last accessed node.
        
        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        self._splay(key)

        # If key is not in the tree, do nothing more.
        if self.root.key != key:
            return

        # Now the node to be deleted is the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
        else:
            # Splay the left subtree to bring its maximum element to its root.
            # We can splay for the key we are deleting, as it's guaranteed to be
            # larger than any key in the left subtree. This brings the max
            # element of the left subtree to its root.
            
            # Temporarily make left_subtree the main tree to splay on it.
            self.root = left_subtree
            self._splay(key)
            
            # The new root (max of left subtree) now has the right subtree attached.
            self.root.right = right_subtree